src/relations/postgresql_client.py

```diff
@@ -0,0 +1,135 @@
+    PostgreSQLDeleteUserError,
+    PostgreSQLGetPostgreSQLVersionError,
+)
+from ops.charm import CharmBase, RelationBrokenEvent
+from ops.framework import Object
+from ops.model import BlockedStatus
+
+from utils import new_password
+
+logger = logging.getLogger(__name__)
+
+
+class PostgreSQLClientProvides(Object):
+    """Defines functionality for the 'provides' side of the 'postgresql-client' relation.
+
+    Hook events observed:
+        - database-requested
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str = "database") -> None:
+        """Constructor for PostgreSQLClientProvides object.
+
+        Args:
+            charm: the charm for which this relation is provided
+            relation_name: the name of the relation
+        """
+        self.relation_name = relation_name
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.charm = charm
+
+        # Charm events defined in the database provides charm library.
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+        self.framework.observe(
+            self.database_provides.on.database_requested, self._on_database_requested
+        )
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
+        """Handle the legacy postgresql-client relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        # Retrieve the database name and extra user roles using the charm library.
+        database = event.database
+        extra_user_roles = event.extra_user_roles
+
+        try:
+            # Creates the user and the database for this specific relation.
+            user = f"relation_id_{event.relation.id}"
+            password = new_password()
+            self.charm.postgresql.create_user(user, password, extra_user_roles=extra_user_roles)
+            self.charm.postgresql.create_database(database, user)
+
+            # Share the credentials with the application.
+            self.database_provides.set_credentials(event.relation.id, user, password)
+
+            # Set the read/write endpoint.
+            self.database_provides.set_endpoints(
+                event.relation.id,
+                f"{self.charm.primary_endpoint}:5432",
+            )
+
+            # Set the read-only endpoint only if there are replicas.
+            if len(self.charm._peers.units) > 0:
```

Comment:
Question: If a replica gets added later on will the `DatabaseRequestedEvent ` get triggered and will this be called?

Reply:
Good point Mia. It was not being handled. I fixed that adding calls to update it on peer relation changed and departed events, and also two new integration tests.

Implemented on https://github.com/canonical/postgresql-k8s-operator/pull/12/commits/79fc87d828b1d6bb996c2055060ed37b2afc55a1.

Reply:
The method from the above commit was improved in https://github.com/canonical/postgresql-k8s-operator/pull/12/commits/188021226e6bec2ef52fb042881ce83ac1ad982e.